File: cpp/3318_find_x_sum/find_x_sum.cpp
#include <cassert>
#include <set>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> findXSum(vector<int>& nums, int k, int x) {
    using pii = pair<int, int>;
    set<pii> l, r;
    int s = 0;
    unordered_map<int, int> cnt;
    auto add = [&](int v) {
      if (cnt[v] == 0) {
        return;
      }
      pii p = {cnt[v], v};
      if (!l.empty() && p > *l.begin()) {
        s += p.first * p.second;
        l.insert(p);
      } else {
        r.insert(p);
      }
    };
    auto remove = [&](int v) {
      if (cnt[v] == 0) {
        return;
      }
      pii p = {cnt[v], v};
      auto it = l.find(p);
      if (it != l.end()) {
        s -= p.first * p.second;
        l.erase(it);
      } else {
        r.erase(p);
      }
    };
    vector<int> ans;
    for (int i = 0; i < nums.size(); ++i) {
      remove(nums[i]);
      ++cnt[nums[i]];
      add(nums[i]);

      int j = i - k + 1;
      if (j < 0) {
        continue;
      }

      while (!r.empty() && l.size() < x) {
        pii p = *r.rbegin();
        s += p.first * p.second;
        r.erase(p);
        l.insert(p);
      }
      while (l.size() > x) {
        pii p = *l.begin();
        s -= p.first * p.second;
        l.erase(p);
        r.insert(p);
      }
      ans.push_back(s);

      remove(nums[j]);
      --cnt[nums[j]];
      add(nums[j]);
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int, int, vector<int>>> tests{
      {{1, 1, 2, 2, 3, 4, 2, 3}, 6, 2, {6, 10, 12}},
      {{3, 8, 7, 8, 7, 5}, 2, 2, {11, 15, 15, 15, 12}},
  };

  for (auto &[nums, k, x, ans] : tests) {
    assert(Solution().findXSum(nums, k, x) == ans);
  }
}

File: rust/3318_find_x_sum/find_x_sum.rs
struct Solution;

impl Solution {
    pub fn find_x_sum(nums: Vec<i32>, k: i32, x: i32) -> Vec<i32> {
        let mut cnt = std::collections::HashMap::new();
        let mut ans = vec![0; nums.len() - k as usize + 1];
        for i in 0..nums.len() {
            *cnt.entry(nums[i]).or_insert(0) += 1;
            if i < k as usize - 1 {
                continue;
            }
            if i >= k as usize {
                if cnt.get(&nums[i - k as usize]).unwrap() == &1 {
                    cnt.remove(&nums[i - k as usize]);
                } else {
                    *cnt.entry(nums[i - k as usize]).or_default() -= 1;
                }
            }
            if cnt.len() <= x as usize {
                for e in &cnt {
                    ans[i - k as usize + 1] += e.0 * e.1;
                }
                continue;
            }
            let mut num_cnts = cnt.iter().collect::<Vec<(&i32, &i32)>>();
            num_cnts.sort_by(|a, b| {
                if a.1 == b.1 {
                    b.0.cmp(&a.0)
                } else {
                    b.1.cmp(&a.1)
                }
            });
            for j in 0..x as usize {
                ans[i - k as usize + 1] += num_cnts[j].0 * num_cnts[j].1;
            }
        }
        ans
    }
}

fn main() {
    let tests = vec![
        (vec![1, 1, 2, 2, 3, 4, 2, 3], 6, 2, vec![6, 10, 12]),
        (vec![3, 8, 7, 8, 7, 5], 2, 2, vec![11, 15, 15, 15, 12]),
    ];

    for (nums, k, x, expected) in tests {
        assert_eq!(Solution::find_x_sum(nums, k, x), expected);
    }
}
